docs/source/en/core_config.rst
Core configuration
==================

This is the place where you can find information about how to configure the
``core`` module of ``sr0wx.py`` project. This is actually my personal
``config.py`` with some additional informations.

What you **won't** find here is how to configure modules. Of course, you can
find it on modules' description pages.

**Warning**: there are some things which I am not proud of. You can probably
find them (and fix them!) on your own. If you will, please send me a patch.

We can divide ``core`` part of ``sr0wx.py`` configuration into several, but
very small, parts:

First, CTCSS (aka subtone). This is going to be easy and self-explanatory: ::

  CTCSS = 88.8
  playCTCSS = False
  CTCSSVolume = 0.1

First line states, that station will be using CTCSS at 88.8 Hz (other possible
values include whole spectrum of floating point numbers or character code from
``'A'`` to ``'AL'``). 

``playCTCSS = False`` means, that CTCSS will be played in
background **only** if any of modules (see below) will ask for it. Other
values include ``True`` (CTCSS is played always) and ``None`` (never).

Next we define how Push-To-Talk is pressed. We do it by sending an RTS signal
(pin 7 on RS232) which is plugged as ``/dev/ttyS0``. ``SerialBaudRate`` is not
used at the moment, but may be usefull in future.::

  serialPort     = '/dev/ttyS0'
  serialBaudRate = 9600
  
As the station is going to speak polish using samples downloaded from Google
Translator we define: (see languages page for mor info) ::
  
  lang = "pl_google"
  
  
Next step is defining how the transmission starts and ends. These are
so-called hello and goodbye messages. As the station is going to give its
callsign in CW we import tiny module to translate text to morse code: ::
 
  from lib.cw import *
  
  helloMsg = ["tu_eksperymentalna_automatyczna_stacja_pogodowa",\
      "sp6yre",cw('sp6yre),]
  goodbyeMsg = ["_","tu_sp6yre",]

Almost done. Defining which modules should be run at each start and in which
order is self explanatory: ::
  
  modules = ["module_a", "module_b"]

Actually, that's all for the ``core``. Next lines (and this is probably 75% of
the whole ``config.py`` are **modules** configuration. This starts from a
dirty hack ...: ::
  
  class m:
      pass

... and usually looks like: ::
  
  foo_module = m()
  foo_module.option_1 = 'bar'
  

Last, but not least, recent versions on ``PyGame`` and/or ``Numeric`` have a
bug which makes CW and CTCSS played twice as fast as it should be and at
doubled frequency. We try to cope with it with: ::

  pygameBug = 1

Oh, almost forget. The values **must** be written in the ``config`` file.


docs/source/pl/core_config.rst
Konfiguracja "rdzenia"
======================

W tym miejscu odnajdziesz informacje na temat jak skonfigurować modłuł główny
(rzeń) ``sr0wx``. W zasadzie zobaczysz fragmenty mojego osobistego pliku
``config.py`` z pewnymi dodatkowymi informacjami. To, czego tu **nie**
znajdziesz to informacje w jaki sposób skonfigurować poszczególne moduły. Te
informacje powinieneś znaleźć na innych stronach tej instrukcji.

**Ostrzeżenie**: w instrukcji (jak i w całym ``sr0wx``) możesz znaleźć rzeczy, z
których nie jestem specjalnie dumny. Będę wdzięczny za przesłanie mi poprawek
(np. w postaci patcha).

Część dotyczącą konfiguracji "rdzenia" ``sr0wx`` można podzielić na kilka,
bardzo niewielkich części:

Po pierwsze, konfiguracja subtonu (dość jasna myślę): ::

  CTCSS = 88.8
  playCTCSS = False
  CTCSSVolume = 0.1

Pierwsza linia oznacza, że stacja używa subtonu 88.8 Hz. Innymi dopuszczalnymi
wartościami są wszystkie liczby zmiennoprzecinkowe jak i kody kanałów do ``'A'``
do ``'AL'``.

Fragment ``playCTCSS = False`` oznacza, że subton CTCSS będzie odtwarzany w tle
**tylko wtedy** gdy któryś z modułów o to poprosi. Innymi dopuszczalnymi wartościami
są ``True`` (ton będzie odtwarzany zawsze) i ``None`` (nigdy).

W następnych liniach mam u siebie zdefiniowane w jaki sposób ``sr0wx.py``
uruchami PTT. Dzieje się to poprzez wysłanie sygnału RTS na pinie 7 RS232
podpiętym jako ``dev/ttyS0``. Wartość ``SerialBaudRate`` nie jest w tym momemcie
w jakikolwiek sposób używana, ale w przyszłości być może. ::

  serialPort     = '/dev/ttyS0'
  serialBaudRate = 9600
  
W związku z tym, że moja stacja mówi po polsku mam zdefiniowany język jako
``pl_google`` (jako, że korzystam z sampli ściągniętych z
``translate.google.com``; zobacz także stronę z innymi językami) ::
  
  lang = "pl_google"
  
W następnym kroku definiujemy treść komunikatu początkowego i końcowego. Jako,
że stacja ma podawać również swój znak telegraficznie ładujemy specjalny moduł
zamieniający tekst na telegrafię: ::
 
  from lib.cw import *
  
  helloMsg = ["tu_eksperymentalna_automatyczna_stacja_pogodowa",\
      "sp6yre",cw('sp6yre),]
  goodbyeMsg = ["_","tu_sp6yre",]

I już prawie koniec. Definiowanie jakie moduły mają być uruchamiane podczas
startu jest równie łatwe: ::
  
  modules = ["module_a", "module_b"]

I to już w zasadzie koniec ustawiania modułu głównego, Kolejne linie
``config.py`` to już tylko konfiguracja poszczególnych modułów. Zaczyna się to
małą sztuczką: ::
  
  class m:
      pass

... i wygląda mniej więcej tak: ::
  
  foo_module = m()
  foo_module.option_1 = 'bar'
  

Na koniec bardzo ważna informacja: współczesne wersje  ``PyGame`` i/lub 
``Numeric`` mają jakiś błąd polegający na tym, że zarówno telegrafia jest
odtwarzana dwa razy szybciej jak powinna (a ton CTCSS ma dwa razy wyższą
częstotliwość. Próbujemy ten błąd obejść poprzez: ::

  pygameBug = 1

Jak mawia klasyk, to by było na tyle (zobacz teraz jak skonfigurować
poszczególne, interesujące Cię moduły)!



config.py
#!/usr/bin/python -tt
# -*- coding: utf-8 -*-

import logging, logging.handlers

log_line_format = '%(asctime)s %(name)s %(levelname)s: %(message)s'
log_handlers = [
    {
        'log_level': logging.DEBUG,
        'class': logging.StreamHandler,
        'config': {'stream': None},
    },
    {
        'log_level': logging.DEBUG,
        'class': logging.handlers.TimedRotatingFileHandler,
        'config': {
            'filename': 'sr0wx.log',
            'when': 'D',
            'interval': 1,
            'backupCount': 30,
            'delay': True,
            'utc': True,
        }
    }
]

CTCSS = 88.8
playCTCSS = False
CTCSSVolume = 0.1
serialPort = '/dev/ttyS0'
serialBaudRate = 9600

from lib.cw import *

lang = "pl_google"

pygameBug = 0

helloMsg = ["tu_eksperymentalna_automatyczna_stacja_pogodowa", "sp6yre", ]
goodbyeMsg = ["_", "tu_sp6yre", cw('sp6yre')]

modules = ["worldweatheronline",]


class m:
    pass

# world weather online

world_weather_online = m()
world_weather_online.api_key = 'CHANGEME'
world_weather_online.latitude = 52.71
world_weather_online.longitude = 19.11
world_weather_online.template = """\
    stan_pogody_z_dnia {OBSERVATION_TIME}
    _ {CURRENT_WEATHER}
    temperatura {CURRENT_TEMP_C} wilgotnosc {CURRENT_HUMIDITY}
    _ kierunek_wiatru {CURRENT_WIND_DIR}
    {CURRENT_WIND_DIR_DEG} predkosc_wiatru {CURRENT_WIND_SPEED_MPS}
    {CURRENT_WIND_SPEED_KMPH} _ cisnienie {CURRENT_PRESSURE}
    pokrywa_chmur {CURRENT_CLOUDCOVER} _

    prognoza_na_nastepne trzy godziny
    {FCAST0_WEATHER} temperatura_minimalna
    {FCAST0_TEMP_MIN_C} maksymalna {FCAST0_TEMP_MAX_C}
    kierunek_wiatru {FCAST0_WIND_DIR} {FCAST0_WIND_DIR_DEG} predkosc_wiatru
    {FCAST0_WIND_SPEED_MPS} {FCAST0_WIND_SPEED_KMPH}

    _ jutro {FCAST1_WEATHER} temperatura_minimalna
    {FCAST1_TEMP_MIN_C} maksymalna {FCAST1_TEMP_MAX_C} kierunek_wiatru
    {FCAST1_WIND_DIR} {FCAST1_WIND_DIR_DEG} predkosc_wiatru
    {FCAST1_WIND_SPEED_MPS} {FCAST1_WIND_SPEED_KMPH} _ """


# -------------
# activity_map
# ------------

activity_map = m()
activity_map.service_url = "http://test.ostol.pl/?base="
activity_map.data = {"callsign": "SR0WX",
                     "lat": 0,
                     "lon": 0,
                     "q": 5,
                     "asl": 118,
                     "agl": 20,
                     "range": 30,
                     "info": u"Additional information",
                     }

